src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt

```diff
@@ -75,80 +79,98 @@ class NexusPublishPlugin : Plugin<Project> {
 
     private fun createRegistry(rootProject: Project): Provider<InvalidatingStagingRepositoryDescriptorRegistry> {
         if (GradleVersion.current() >= GradleVersion.version("6.1")) {
-            return rootProject.gradle.sharedServices.registerIfAbsent("stagingRepositoryUrlRegistry", StagingRepositoryDescriptorRegistryBuildService::class.java) {}.map { it.registry }
+            return rootProject.gradle.sharedServices.registerIfAbsent(
+                "stagingRepositoryUrlRegistry",
+                StagingRepositoryDescriptorRegistryBuildService::class,
+                Action { }
+            ).map { it.registry }
         }
         val registry = InvalidatingStagingRepositoryDescriptorRegistry()
         return rootProject.provider { registry }
     }
 
     private fun configureNexusTasks(rootProject: Project, extension: NexusPublishExtension, registry: Provider<InvalidatingStagingRepositoryDescriptorRegistry>) {
         rootProject.tasks.withType(AbstractNexusStagingRepositoryTask::class.java).configureEach {
-            clientTimeout.convention(extension.clientTimeout)
-            connectTimeout.convention(extension.connectTimeout)
-            repositoryDescription.convention(extension.repositoryDescription)
-            useStaging.convention(extension.useStaging)
+            this.clientTimeout.convention(extension.clientTimeout)
+            this.connectTimeout.convention(extension.connectTimeout)
+            this.repositoryDescription.convention(extension.repositoryDescription)
+            this.useStaging.convention(extension.useStaging)
+            //this.repository.convention() is set inside extension.repositories.all { }.
         }
         rootProject.tasks.withType(AbstractTransitionNexusStagingRepositoryTask::class.java).configureEach {
-            transitionCheckOptions.convention(extension.transitionCheckOptions)
+            this.transitionCheckOptions.convention(extension.transitionCheckOptions)
+            this.stagingRepositoryId.convention(registry.map { it[repository.get().name].stagingRepositoryId })
         }
         extension.repositories.all {
-            username.convention(rootProject.provider { rootProject.findProperty("${name}Username") as? String })
-            password.convention(rootProject.provider { rootProject.findProperty("${name}Password") as? String })
-            publicationType.convention(PublicationType.MAVEN)
-
             val repository = this
-            val retrieveStagingProfileTask = rootProject.tasks.register<RetrieveStagingProfile>("retrieve${capitalizedName}StagingProfile", rootProject.objects, extension, repository)
+
+            repository.username.convention(rootProject.provider { rootProject.findProperty("${name}Username") as? String })
+            repository.password.convention(rootProject.provider { rootProject.findProperty("${name}Password") as? String })
```

Comment:
@szpak I wanted to replace this with:
```suggestion
            repository.username.convention(rootProject.gradleProperty("${name}Username"))
            repository.password.convention(rootProject.gradleProperty("${name}Password"))
```
so badly, but that method is only avaiable from Gradle 6.2.

Also it would be an inconvenient breaking change, because right now `findProperty` does a ton of things (see [javadoc](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#findProperty-java.lang.String-)), while `.gradleProperty` is just `-P` on the command line or `gradle.properties`. I think either it needs to be standardized `-P` or `-D`, or use something like:
```kotlin
/**
 * @see Project.findProperty as it was used to find this, but it's impractical do the same in the new lazy API.
 */
private fun ProviderFactory.findProperty(propertyName: String): Provider<String> =
    gradleProperty(propertyName).orElse(systemProperty(propertyName))
```
The rest of the steps in `findProperty` are either deprecated or totally magical.

Reply:
Hmm, are you sure that `findProperty` searches also for system properties (`-D`)? Definitely, `-P` was intended to use here (in GNPP).

I wonder, how many people could use the other cases mentioned in javadoc?

Btw, it's misleading - in code review in a browser - to distinguish `orElse(T)` and `orElse(Provider<T>)`. I like a [name separation](https://github.com/gradle/gradle/issues/10520) in Java. However, designing API is hard :-).

Btw2, a fun fact is that I proposed and implemented the initial version of `findProperty` in Gradle years ago (I use so often the tertiary operator in `build.gradle`...). However, the truth is, the aforementioned "magic" it/was itself already implemented at the lower level - https://github.com/gradle/gradle/pull/578/files#diff-2b0a49af16e081b46065d37c7fe53d37b8ee2fc036d9cdbe613f603bfd4e1ef9R778

Reply:
> are you sure that findProperty searches also for system properties (-D)

huh, you're right, that's good then, `providers.gradleProperty(name)` would be enough. Tried `println(findProperty("foo"))`:
 * `gradlew` &rarr; null
 * `gradlew -Pfoo=bar` &rarr; "bar"
 * `gradlew -Dfoo=bar` &rarr; null
 * `gradle.properties` `foo=bar` &rarr; "bar"
 * `gradle.properties` `systemProp.foo=bar` &rarr; null

The limitation still applies: min is Gradle 6.0 now, and the method is only from 6.2.

---

> I wonder, how many people could use the other cases mentioned in javadoc?

The only one I can imagine is (4) `ext.sonatypeUsername` like [this](https://github.com/square/retrofit/commit/4fa58ebb9781502d9afc1ac3c6273b117a667e76#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7), but then the username/password is hardcoded in plain sight. (3) is deprecated, but (2) and (3) could be used to propagate a value from a plugin, like
```
target.extensions.create("sonatypePassword", stringRetrivedFromUnthinkableSource)
```
if that's even valid.

---

> Btw, it's misleading - in code review in a browser - to distinguish orElse(T) and orElse(Provider<T>)

I agree the problem is not with orElse, but with Kotlin's implicit receivers IMO:
```kotlin
private fun ProviderFactory.findProperty(propertyName: String): Provider<String> =
    this.gradleProperty(propertyName).orElse(this.systemProperty(propertyName))
```
in this case it's clear(er) that both of them are providers. and that `systemProperty` is not a theoretical `project.systemProperty(name): String`.

---

> Btw2, a fun fact is that I proposed and implemented the initial version of findProperty

Haha, nice! 👏 reminds me: "designing API is hard" ... `project.propertyOrNull(name)` would be the name according to Kotlin stdlib conventions, which I think are really convenient.

---

Btw3, when I first came to this repo, I knew your name was familiar from somewhere... know I know, it's [PIT](https://github.com/szpak/gradle-pitest-plugin)!

Reply:
> The limitation still applies: min is Gradle 6.0 now, and the method is only from 6.2.

I see no problem. People using 6.0 should be able to upgrade to 6.2 without any issues, so let's bump the minimal Gradle version to 6.2.

Reply:
Let's leave it as is for this PR, I might open another one to see what a 6.2 bump would look like.

Reply:
Sound good.

Reply:
Opened https://github.com/TWiStErRob/nexus-publish-gradle-plugin/pull/7 for now until this PR gets merged.